src/dmd/expression.d

```diff
@@ -1308,6 +1214,17 @@ extern (C++) abstract class Expression : ASTNode
         return err;
     }
 
+    /*
+    Check if sc.func is impure or can be made impure.
+    Returns true on error, i.e. if sc.func is pure and cannot be made impure.
```

Comment:
@aG0aep6G @thewilsonator when writing function comments, please use the Ddoc form.

Also, the function is not checking for an error, and it does more than check - it sets an impure state.

The phrase "cannot be made impure" makes little sense, as a pure function can always be called by an impure one.

Reply:
> Also, the function is not checking for an error, and it does more than check - it sets an impure state.

How about:
```d
    /*********************************************
     * Try marking sc.func as impure.
     * Returns true if sc.func is pure and cannot be made impure.
     */
    private static bool tryMakingImpure(Scope* sc) { ... }
```

> The phrase "cannot be made impure" makes little sense, as a pure function can always be called by an impure one.

This is about the case where a (pure) function tries calling an impure one. `sc.func` is the calling function, and we try to mark it as impure. But that might fail (because of an explicit `pure` attribute). In that case, `sc.func` "cannot be made impure". Makes sense?

Reply:
Or maybe this name and comment:

```d
    /*********************************************
     * An impure function is called in sc. Propagate the impurity to the
     * surrounding function (sc.func).
     * Returns true if sc.func is pure and cannot be made impure.
     */
    private static bool propagateImpure(Scope* sc)
```